examples/voice-assistant/log_transcriptions.py

```diff
@@ -0,0 +1,78 @@
+import asyncio
+from datetime import datetime
+
+from aiofile import async_open as open
+from dotenv import load_dotenv
+from livekit import rtc
+from livekit.agents import AutoSubscribe, JobContext, WorkerOptions, cli, llm
+from livekit.agents.voice_assistant import VoiceAssistant
+from livekit.plugins import deepgram, openai, silero
+
+load_dotenv()
+
+
+async def entrypoint(ctx: JobContext):
+    initial_ctx = llm.ChatContext().append(
+        role="system",
+        text=(
+            "You are a voice assistant created by LiveKit. Your interface with users will be voice. "
+            "You should use short and concise responses, and avoiding usage of unpronouncable punctuation."
+        ),
+    )
+
+    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)
+
+    assistant = VoiceAssistant(
+        vad=silero.VAD.load(),
+        stt=deepgram.STT(),
+        llm=openai.LLM(),
+        tts=openai.TTS(),
+        chat_ctx=initial_ctx,
+    )
+    assistant.start(ctx.room)
+
+    # listen to incoming chat messages, only required if you'd like the agent to
+    # answer incoming messages from Chat
+    chat = rtc.ChatManager(ctx.room)
+
+    async def answer_from_text(txt: str):
+        chat_ctx = assistant.chat_ctx.copy()
+        chat_ctx.append(role="user", text=txt)
+        stream = assistant.llm.chat(chat_ctx=chat_ctx)
+        await assistant.say(stream)
+
+    @chat.on("message_received")
+    def on_chat_received(msg: rtc.ChatMessage):
+        if msg.message:
+            asyncio.create_task(answer_from_text(msg.message))
+
+    # clear file before writing to it
+    async with open("transcriptions.log", "w"):
+        pass
+
+    async def log_user_speech(msg: llm.ChatMessage):
+        # convert string lists to strings, drop images
+        if isinstance(msg.content, list):
+            msg.content = "\n".join(
+                "[image]" if isinstance(x, llm.ChatImage) else x for x in msg
+            )
+        async with open("transcriptions.log", "a+") as f:
+            await f.write(f"[{datetime.now()}] USER:\n{msg.content}\n\n")
+
+    async def log_agent_speech(msg: llm.ChatMessage):
+        async with open("transcriptions.log", "a+") as f:
```

Comment:
what about using a queue here instead of opening the file each time?

IMO this would be a cleaner design:
* log_user_speech and log_agent_speech push a dict with details to a asyncio.Queue
* spin up only a single task to write to file
* the task should flush periodically and close itself when disconnected
* we can avoid creating separate tasks for these callbacks due to synchronization issues (i.e. if both things write at the same time, chatlog is corrupted)

Reply:
sure, but this isn't production code. it shows that the agent and user texts are reachable like *this* and then you can do whatever with them: log to a file, append to a database, forward to somewhere else, etc. minimal_assistant isn't a demo you're gonna run on prod, having a lot of syscalls doesn't matter here imo